app/src/verifier.h

```diff
@@ -66,76 +58,322 @@ namespace scitt::verifier
+      {
+        throw cose::COSEDecodeError(
+          "cty must be 'application/vnd.cncf.notary.payload.v1+json' for "
+          "Notary claims");
+      }
+
+      if (notary_signing_scheme.value() == "notary.x509")
+      {
+        // notary_signing_time is not critical but is required iff notary.x509
+        if (!notary_signing_time.has_value())
         {
-          jwk = did::find_assertion_method_jwk_in_did_document(
-            resolution.did_doc, assertion_method_id);
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.signingTime in protected header");
         }
-        catch (const did::DIDAssertionMethodError& e)
+        if (notary_authentic_signing_time.has_value())
         {
-          throw VerificationError(e.what());
+          throw cose::COSEDecodeError(
+            "io.cncf.notary.authenticSigningTime not allowed in protected "
+            "header when io.cncf.notary.signingScheme is `notary.x509`");
+        }
+      }
+      else if (notary_signing_scheme.value() == "notary.x509.signingAuthority")
+      {
+        // notary_authentic_signing_time is critical and required iff
+        // notary.x509.signingAuthority
+        if (notary_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "Notary io.cncf.notary.signingTime not allowed in protected header "
+            "when io.cncf.notary.signingScheme is "
+            "`notary.x509.signingAuthority`");
+        }
+        if (!notary_authentic_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.authenticSigningTime in protected header");
+        }
+        if (!phdr.is_critical("io.cncf.notary.authenticSigningTime"))
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.authenticSigningTime in crit parameters");
         }
+      }
+      else
+      {
+        throw cose::COSEDecodeError(
+          "Notary io.cncf.notary.signingScheme must be `notary.x509` or "
+          "`notary.x509.signingAuthority`");
+      }
 
-        // Convert the JWK into something we can actually use.
-        key = get_jwk_public_key(jwk);
+      if (notary_expiry.has_value())
+      {
+        // notary_expiry is critial but not required.
+        if (!phdr.is_critical("io.cncf.notary.expiry"))
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.expiry in crit parameters");
+        }
       }
+    }
+
+    PublicKey process_notary_profile(
+      cose::UnprotectedHeader& uhdr,
+      cose::ProtectedHeader& phdr,
+      kv::Tx& tx,
+      const Configuration& configuration)
+    {
+      // Validate protected header
+      validate_notary_protected_header(phdr, configuration);
+
+      // Validate unprotected header
+      // Unprotected header must contain x5chain else the notary claim cannot be
+      // verified.
+      if (!uhdr.x5chain.has_value())
```

Comment:
[The Notary COSE spec](https://github.com/notaryproject/notaryproject/blob/900a36df30315803e221bd45754a53395fae359b/specs/signature-envelope-cose.md#unprotected-headers) says `Optionally, this header can be presented in the protected header`, which suggests we should actually look at both headers. Probably we should be looking at the protected one first, or even fail if both are present.

(oddly enough, the JWS spec has no such phrasing, and only allows it in the unprotected header)

Reply:
Done in https://github.com/microsoft/scitt-ccf-ledger/pull/90